Failed to load ApplicationContext after update to 1.4.0

I use spring boot 2.2.5-RELEASE, after update a micrometer-registry-new-relic from 1.3.5 to 1.4.0 my application does not start.

Stacktrace:
```
Caused by: org.springframework.beans.factory.BeanCreationException: Error creating bean with name 'newRelicMeterRegistry' defined in class path resource [org/springframework/boot/actuate/autoconfigure/metrics/export/newrelic/NewRelicMetricsExportAutoConfiguration.class]: Bean instantiation via factory method failed; nested exception is org.springframework.beans.BeanInstantiationException: Failed to instantiate [io.micrometer.newrelic.NewRelicMeterRegistry]: Factory method 'newRelicMeterRegistry' threw exception; nested exception is java.lang.NoSuchMethodError: 'io.micrometer.newrelic.NewRelicMeterRegistry$Builder io.micrometer.newrelic.NewRelicMeterRegistry$Builder.httpClient(io.micrometer.core.ipc.http.HttpSender)'
	at org.springframework.beans.factory.support.ConstructorResolver.instantiate(ConstructorResolver.java:656)
	at org.springframework.beans.factory.support.ConstructorResolver.instantiateUsingFactoryMethod(ConstructorResolver.java:636)
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.instantiateUsingFactoryMethod(AbstractAutowireCapableBeanFactory.java:1338)
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.createBeanInstance(AbstractAutowireCapableBeanFactory.java:1177)
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.doCreateBean(AbstractAutowireCapableBeanFactory.java:557)
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.createBean(AbstractAutowireCapableBeanFactory.java:517)
	at org.springframework.beans.factory.support.AbstractBeanFactory.lambda$doGetBean$0(AbstractBeanFactory.java:323)
	at org.springframework.beans.factory.support.DefaultSingletonBeanRegistry.getSingleton(DefaultSingletonBeanRegistry.java:222)
	at org.springframework.beans.factory.support.AbstractBeanFactory.doGetBean(AbstractBeanFactory.java:321)
	at org.springframework.beans.factory.support.AbstractBeanFactory.getBean(AbstractBeanFactory.java:202)
	at org.springframework.beans.factory.config.DependencyDescriptor.resolveCandidate(DependencyDescriptor.java:276)
	at org.springframework.beans.factory.support.DefaultListableBeanFactory.doResolveDependency(DefaultListableBeanFactory.java:1287)
	at org.springframework.beans.factory.support.DefaultListableBeanFactory.resolveDependency(DefaultListableBeanFactory.java:1207)
	at org.springframework.beans.factory.support.ConstructorResolver.resolveAutowiredArgument(ConstructorResolver.java:885)
	at org.springframework.beans.factory.support.ConstructorResolver.createArgumentArray(ConstructorResolver.java:789)
	... 193 more
Caused by: org.springframework.beans.BeanInstantiationException: Failed to instantiate [io.micrometer.newrelic.NewRelicMeterRegistry]: Factory method 'newRelicMeterRegistry' threw exception; nested exception is java.lang.NoSuchMethodError: 'io.micrometer.newrelic.NewRelicMeterRegistry$Builder io.micrometer.newrelic.NewRelicMeterRegistry$Builder.httpClient(io.micrometer.core.ipc.http.HttpSender)'
	at org.springframework.beans.factory.support.SimpleInstantiationStrategy.instantiate(SimpleInstantiationStrategy.java:185)
	at org.springframework.beans.factory.support.ConstructorResolver.instantiate(ConstructorResolver.java:651)
	... 207 more
Caused by: java.lang.NoSuchMethodError: 'io.micrometer.newrelic.NewRelicMeterRegistry$Builder io.micrometer.newrelic.NewRelicMeterRegistry$Builder.httpClient(io.micrometer.core.ipc.http.HttpSender)'
	at org.springframework.boot.actuate.autoconfigure.metrics.export.newrelic.NewRelicMetricsExportAutoConfiguration.newRelicMeterRegistry(NewRelicMetricsExportAutoConfiguration.java:71)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.springframework.beans.factory.support.SimpleInstantiationStrategy.instantiate(SimpleInstantiationStrategy.java:154)
	... 208 more
```
It seems that you removed the method which is used by spring boot in autoconfiguration for plugin. Backward compatibility is broken now:
`NewRelicMetricsExportAutoConfiguration`:
```java
@Bean
  @ConditionalOnMissingBean
  public NewRelicMeterRegistry newRelicMeterRegistry(NewRelicConfig newRelicConfig, Clock clock) {
    return NewRelicMeterRegistry.builder(newRelicConfig).clock(clock).httpClient(new HttpUrlConnectionSender(this.properties.getConnectTimeout(), this.properties.getReadTimeout())).build();
  }
```

--- a/implementations/micrometer-registry-new-relic/src/main/java/io/micrometer/newrelic/NewRelicInsightsApiClientProvider.java
+++ b/implementations/micrometer-registry-new-relic/src/main/java/io/micrometer/newrelic/NewRelicInsightsApiClientProvider.java
@@ -62,7 +62,8 @@ public class NewRelicInsightsApiClientProvider implements NewRelicClientProvider
     private final Logger logger = LoggerFactory.getLogger(NewRelicInsightsApiClientProvider.class);
 
     private final NewRelicConfig config;
-    private final HttpSender httpClient;
+    // VisibleForTesting
+    final HttpSender httpClient;
     private final NamingConvention namingConvention;
     private final String insightsEndpoint;
 

--- a/implementations/micrometer-registry-new-relic/src/main/java/io/micrometer/newrelic/NewRelicMeterRegistry.java
+++ b/implementations/micrometer-registry-new-relic/src/main/java/io/micrometer/newrelic/NewRelicMeterRegistry.java
@@ -22,6 +22,8 @@ import io.micrometer.core.instrument.Clock;
 import io.micrometer.core.instrument.config.NamingConvention;
 import io.micrometer.core.instrument.step.StepMeterRegistry;
 import io.micrometer.core.instrument.util.NamedThreadFactory;
+import io.micrometer.core.ipc.http.HttpSender;
+import io.micrometer.core.lang.Nullable;
 
 /**
  * Publishes metrics to New Relic Insights based on client provider selected (API or Java Agent).
@@ -58,7 +60,7 @@ public class NewRelicMeterRegistry extends StepMeterRegistry {
     }
 
     // VisibleForTesting
-    NewRelicMeterRegistry(NewRelicConfig config, NewRelicClientProvider clientProvider,
+    NewRelicMeterRegistry(NewRelicConfig config, @Nullable NewRelicClientProvider clientProvider,
                 NamingConvention namingConvention, Clock clock, ThreadFactory threadFactory) {
         super(config, clock);
 
@@ -93,17 +95,20 @@ public class NewRelicMeterRegistry extends StepMeterRegistry {
     public static class Builder {
         private final NewRelicConfig config;
 
+        @Nullable
         private NewRelicClientProvider clientProvider;
         private NamingConvention convention = new NewRelicNamingConvention();
         private Clock clock = Clock.SYSTEM;
         private ThreadFactory threadFactory = DEFAULT_THREAD_FACTORY;
+        @Nullable
+        private HttpSender httpClient;
 
         Builder(NewRelicConfig config) {
             this.config = config;
         }
 
         /**
-         * Use the client provider.
+         * Use the client provider. This will override {@link NewRelicConfig#clientProviderType()}.
          * @param clientProvider client provider to use
          * @return builder
          * @since 1.4.0
@@ -134,7 +139,26 @@ public class NewRelicMeterRegistry extends StepMeterRegistry {
             return this;
         }
 
+        /**
+         * Note: This only has an effect when a {@link #clientProvider(NewRelicClientProvider)} is not provided and {@link NewRelicConfig#clientProviderType()} is {@link ClientProviderType#INSIGHTS_API}.
+         *
+         * @param httpClient http client to use for publishing
+         * @return builder
+         * @deprecated use {@link #clientProvider(NewRelicClientProvider)} instead.
+         */
+        @Deprecated
+        public Builder httpClient(HttpSender httpClient) {
+            this.httpClient = httpClient;
+            return this;
+        }
+
         public NewRelicMeterRegistry build() {
+            if (clientProvider == null && httpClient != null) {
+                //default to Insight API client provider if not specified in config or provided
+                clientProvider = (config.clientProviderType() == ClientProviderType.INSIGHTS_AGENT)
+                        ? new NewRelicInsightsAgentClientProvider(config)
+                        : new NewRelicInsightsApiClientProvider(config, httpClient, convention);
+            }
             return new NewRelicMeterRegistry(config, clientProvider, convention, clock, threadFactory);
         }
     }

--- a/implementations/micrometer-registry-new-relic/src/test/java/io/micrometer/newrelic/NewRelicMeterRegistryTest.java
+++ b/implementations/micrometer-registry-new-relic/src/test/java/io/micrometer/newrelic/NewRelicMeterRegistryTest.java
@@ -925,6 +925,24 @@ class NewRelicMeterRegistryTest {
 
         assertThat(getInsightsAgentClientProvider(config)).isNotNull();
     }
+
+    @Test
+    void canCustomizeHttpSenderViaBuilder_deprecated() {
+        HttpSender httpSender = mock(HttpSender.class);
+        NewRelicClientProvider clientProvider = NewRelicMeterRegistry.builder(insightsApiConfig).httpClient(httpSender).build().clientProvider;
+        assertThat(clientProvider).isInstanceOf(NewRelicInsightsApiClientProvider.class);
+        assertThat(((NewRelicInsightsApiClientProvider) clientProvider).httpClient).isEqualTo(httpSender);
+    }
+
+    @Test
+    void canCustomizeHttpSenderViaBuilder() {
+        HttpSender httpSender = mock(HttpSender.class);
+        NewRelicClientProvider clientProvider = NewRelicMeterRegistry.builder(insightsApiConfig)
+                .clientProvider(new NewRelicInsightsApiClientProvider(insightsApiConfig, httpSender, new NewRelicNamingConvention()))
+                .build().clientProvider;
+        assertThat(clientProvider).isInstanceOf(NewRelicInsightsApiClientProvider.class);
+        assertThat(((NewRelicInsightsApiClientProvider) clientProvider).httpClient).isEqualTo(httpSender);
+    }
     
     static class MockHttpSender implements HttpSender {
         

